aptos-move/aptos-vm/src/move_vm_ext/vm.rs

```diff
@@ -44,6 +44,10 @@ impl MoveVmExt {
         extensions.add(NativeTransactionContext::new(script_hash));
         extensions.add(NativeCodeContext::default());
 
+        // The VM code loader has bugs around module upgrade. After a module upgrade, the internal
+        // cache needs to be flushed to work around those bugs.
+        self.inner.flush_loader_cache_if_invalidated();
```

Comment:
do we mark invalidated anywhere?

Reply:
The VM does it itself when it sees a module upgrade. I actually wanted to do the flush first also internally to the VM, but it is not working because the assumption is you only have one session by VM at a given time. (I think AptosVM satisfies this.) In any case, if you have an outside reason to invalidate, there is an API for that. The outside reason can be e.g. that you abandon a module publish and don't commit it to storage, so you kind of need to delete it from the cache.